### DISCUSSION TITLE
리뷰 수정 시 사진을 처리하는 방법

### DISCUSSION THREAD

**Discussion by Kim0914** [Busan-Dinosaur/Foodbowl-Backend#165] [의논해볼까요?]
## 서론

현재 비즈니스 요구사항 중 리뷰 수정이 존재합니다.

리뷰 수정 기능에서는 `리뷰 사진`, `리뷰 내용`을 수정할 수 있습니다.
우선 클라이언트가 가게에 대한 리뷰를 조회했을 때, 리뷰 하나가 가지고 있는 값은 아래와 같이 예상할 수 있습니다.

```json
{
    "id": 1,
    "storeId": 5,
    "content": "맛있지만, 가격은 살짝 비싸요.",
    "photos": {
                  {
                      "id": 20,
                      "path": "https://justdoeat.shop/static/images/5/image1.png"
                  },
                  {
                      "id": 21,
                      "path": "https://justdoeat.shop/static/images/5/image2.png"
                  },
                  {
                      "id": 22,
                      "path": "https://justdoeat.shop/static/images/5/image3.png"
                  }
            },

    ....생략....
}

```

## 본론

리뷰를 수정하는 기능에서 가장 핵심적인 부분이 `삭제하려는 사진을 어떻게 삭제하고, 추가하려는 사진을 어떻게 추가할 지`라고 생각합니다 !

현재 떠오르는 방법들은 다음과 같습니다.

### 1. 수정 시 삭제하려는 사진 ID와 새롭게 추가될 사진 데이터 받기

클라이언트로 부터 `삭제가 필요한 사진의 ID`와 `새롭게 추가될 사진들`을 요청으로 받습니다.

1. `현재 사진 개수 - 삭제하는 사진 개수 + 추가될 사진 개수`가 4장 이하인지 확인
2. 삭제하려는 사진 엔티티가 리뷰에 포함된 사진인지 확인
3. 해당 사진 삭제하기
4. 새로운 사진 추가하기

이 방법은 ID(PK) 기반으로 조회, 삭제가 이뤄질 것 같아 직관적인 방법이라고 생각합니다.

조금 걸리는 점은, 수정 시 클라이언트가 보내는 요청 형태가 까다로울 수 있다는 점이라고 생각합니다.
(크게 까다롭지 않을 것이라는 생각도 들어요 ㅎㅎ)

### 2. 수정 요청에서 발생하는 이미지와 수정 전 이미지 비교하기

클라이언트는 수정 폼에 있는 이미지를 그대로 전송합니다. 

1. 수정 전 리뷰가 가지고 있던 이미지들을 DB에서 가져온다.
2. 수정 요청으로 들어온 MultipatFile(이미지)의 이름을 가져온다.
3. 수정 요청으로 들어온 이미지가 수정 전 리뷰 이미지에 존재하면 pass
4. 없으면 생성할 리스트에 추가
5. 기존 이미지들에서 삭제가 필요한 이미지 분리
6. 이미지 개수 검증
7. 추가할 이미지 생성, 삭제할 이미지 삭제

이 방법의 장점으로는 클라이언트가 요청을 보낼 때 삭제를 고려하지 않고 보낼 수 있다는 점..? 정도가 생각나네요.

다만 이미지를 저장할 때 파일 명을 UUID를 사용하지 않고 특정 해시 값을 이용해 디코딩할 수 있도록 인코딩을 해야할 것 같습니다 !
(이미지 originalName에서 파일을 찾을 수 있도록)

단점으로는 인코딩/디코딩을 하는 개발적인 비용과 동일한 이름을 가진 이미지가 들어왔을 때 추가적으로 처리해줘야합니다.

생각해본 방법은 이정도 떠오르는데 또다른 아이디어 있으시면 남겨주십쇼 🤓

**Comment by woo-chang** [comment]
각각의 장단점을 잘 비교해주셨네요 ㅎㅎ

`DB`에 저장되어 있는 데이터인만큼 `PK`를 사용하는 것이 직관적이고 안정적으로 데이터를 관리할 수 있을 것 같아요. 추가적으로 `PK`는 유니크한 값이지만, 이미지명은 중복될 수 있는 값이기 때문에 해당 방법으로 관리하게 된다면 이를 관리하는 추가적인 리소스도 발생할 것 같다는 생각이 들었어요!

따라서 같이 얘기한대로 1번 방법으로 진행하면 좋을 것 같아요 🤓

추가적으로는 이미지 저장, 삭제를 하나의 트랜잭션에서 어떻게 분리할 수 있을지, 백그라운드에서 저장, 삭제가 일어나도록 할 수 있을지, 트랜잭션 롤백 시 이미지는 어떻게 롤백시킬 수 있을지 고민해보면 좋을 것 같아요 :)

**Reply by Kim0914** [reply]
굿굿 좋아용
